### ISSUE TITLE
paddle1.3.2，使用reader.multiprocess_reader读数训练时，偶发性卡住

### ISSUE THREAD

**Issue by xiangyubo** [PaddlePaddle/Paddle#16961]
1）PaddlePaddle版本：1.3.2
    3）GPU：v100
    4）系统环境：python2.7，AI Studio
- 训练信息
    1）单机
- 问题描述：在训练 YOLOv3 模型的时候，会出现偶发性的卡住，AI Studio 中也有其他用户反馈训练中卡住的问题。
训练中，使用的是自定义的数据集，423张图片。使用 reader.multiprocess_reader 做异步读取。读取部分定义：

def multi_process_custom_reader(file_path, data_dir, num_workers, mode):
    file_path = os.path.join(data_dir, file_path)
    readers = []
    images = [line.strip() for line in open(file_path)]
    n = int(math.ceil(len(images) // num_workers))
    image_lists = [images[i: i + n] for i in range(0, len(images), n)]
    for l in image_lists:
        readers.append(paddle.batch(custom_reader(l, data_dir, mode),
                                    batch_size=train_parameters['train_batch_size'],
                                    drop_last=True))
    return paddle.reader.multiprocess_reader(readers, False)

训练部分：
for pass_id in range(train_parameters["num_epochs"]):
    logger.info("current pass: %d, start read image", pass_id)
    batch_id = 0
    train_reader.start()
    try:
        while True:
            t1 = time.time()
            loss = exe.run(train_program, fetch_list=train_fetch_list)
            period = time.time() - t1
            loss = np.mean(np.array(loss))
            batch_id += 1
            total_batch_count += 1
            # balabala....一顿操作
    except fluid.core.EOFException:
        train_reader.reset()

**Comment by shippingwang** [comment]
> 会出现偶发性的卡住 

@xiangyubo

这个您是指可以读数据，但是训到一半卡住了
还是说一开始就读不了呢？

**Comment by xiangyubo** [comment]
> > 会出现偶发性的卡住
> 
> @xiangyubo
> 
> 这个您是指可以读数据，但是训到一半卡住了
> 还是说一开始就读不了呢？

@shippingwang 
训练到一半卡住了

**Comment by wangguibao** [comment]
用户在AI studio上执行训练。后来怀疑是当开的reader进程过多时会有调度问题。将reader进程从8个改为4个后，继续观察。至今天未再反馈问题。

**Comment by xiangyubo** [comment]
> 用户在AI studio上执行训练。后来怀疑是当开的reader进程过多时会有调度问题。将reader进程从8个改为4个后，继续观察。至今天未再反馈问题。

改到1、2的时候不会卡住，大了会卡住。

**Comment by xiangyubo** [comment]
我现在使用 
paddle 1.4.1 
python 3.6 
GPU，v100
batch_size：128 、64、32、16
reader 个数：4、2、1
训练 yolo 的时候，也使用上述相似数据读取方式，我打印了读数的日志。数据读取没有卡住，但是训练的部分 exe.run() 会卡住，这次是从一开始就卡住。我同时读数的 reader 进程数量从4到2到1，都会卡住。
我把 batch_size 调到 8 的时候不会卡住，此时的 reader 是1个。看到的现象是数据匀速读取，但是 exe.run 的输出是批量式的出结果。批量出8次结果，第一次 exe.run 会消耗几十秒，后面的都只消耗 0.04 秒左右。
=========================分割线================================
刚刚我测试了一下等久一些，batch_size 128，reader 4个，能训练，但是第一次 exe.run 出来的结果要等1800+秒，很容易让人觉得 exe.run 卡住了。但是跑了32个 batch 之后似乎不读数据了
